File: backend/app/models/event.py
import uuid
from datetime import datetime
from enum import Enum

from sqlalchemy import String, Boolean, Float, DateTime, ForeignKey
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin


class EventStatus(str, Enum):
    SCHEDULED = "scheduled"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    SKIPPED = "skipped"


class Event(Base, TimestampMixin):
    __tablename__ = "events"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    task_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tasks.id", ondelete="SET NULL")
    )
    schedule_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("schedules.id", ondelete="SET NULL")
    )

    title: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str | None] = mapped_column(String(2000))
    location: Mapped[str | None] = mapped_column(String(500))

    start_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)
    end_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)
    is_all_day: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    status: Mapped[str] = mapped_column(String(20), default=EventStatus.SCHEDULED, nullable=False)
    is_agent_created: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    is_locked: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    # External calendar provider info
    provider: Mapped[str | None] = mapped_column(String(50))
    provider_event_id: Mapped[str | None] = mapped_column(String(255))

    # Actual timing (filled in by activity tracking)
    actual_start_time: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    actual_end_time: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    completion_percentage: Mapped[float] = mapped_column(Float, default=0.0, nullable=False)

    # Provider-specific extra fields
    metadata_: Mapped[dict | None] = mapped_column("metadata", JSONB)

    user = relationship("User", back_populates="events")
    task = relationship("Task", back_populates="events")
    schedule = relationship("Schedule", back_populates="events")
    activity_logs = relationship("ActivityLog", back_populates="event")

File: mobile/lib/models/event.dart
class Event {
  final String id;
  final String userId;
  final String? taskId;
  final String? scheduleId;
  final String title;
  final String? description;
  final String? location;
  final DateTime startTime;
  final DateTime endTime;
  final bool isAllDay;
  final String status;
  final bool isAgentCreated;
  final bool isLocked;
  final double completionPercentage;
  final String? provider;

  const Event({
    required this.id,
    required this.userId,
    this.taskId,
    this.scheduleId,
    required this.title,
    this.description,
    this.location,
    required this.startTime,
    required this.endTime,
    required this.isAllDay,
    required this.status,
    required this.isAgentCreated,
    required this.isLocked,
    required this.completionPercentage,
    this.provider,
  });

  factory Event.fromJson(Map<String, dynamic> json) => Event(
        id: json['id'] as String,
        userId: json['user_id'] as String,
        taskId: json['task_id'] as String?,
        scheduleId: json['schedule_id'] as String?,
        title: json['title'] as String,
        description: json['description'] as String?,
        location: json['location'] as String?,
        startTime: DateTime.parse(json['start_time'] as String),
        endTime: DateTime.parse(json['end_time'] as String),
        isAllDay: json['is_all_day'] as bool,
        status: json['status'] as String,
        isAgentCreated: json['is_agent_created'] as bool,
        isLocked: json['is_locked'] as bool,
        completionPercentage: (json['completion_percentage'] as num).toDouble(),
        provider: json['provider'] as String?,
      );
}
